File: ddss/egg.ts
import { Rule } from "atsds";
import { Search } from "./egraph.js";
import { initializeDatabase, insertOrIgnore, Fact, Idea } from "./orm.js";
import { type Sequelize, Op } from "sequelize";

export async function main(addr: string, sequelize?: Sequelize) {
    if (!sequelize) {
        sequelize = await initializeDatabase(addr);
    }

    try {
        const search = new Search();
        let pool: Rule[] = [];
        let maxFact = -1;
        let maxIdea = -1;

        while (true) {
            const begin = Date.now();
            let count = 0;

            const newIdeas = await Idea.findAll({
                where: { id: { [Op.gt]: maxIdea } },
                order: [["id", "ASC"]],
            });
            for (const idea of newIdeas) {
                maxIdea = Math.max(maxIdea, idea.id);
                pool.push(new Rule(idea.data));
            }

            const newFacts = await Fact.findAll({
                where: { id: { [Op.gt]: maxFact } },
                order: [["id", "ASC"]],
            });
            for (const fact of newFacts) {
                maxFact = Math.max(maxFact, fact.id);
                search.add(new Rule(fact.data));
            }

            search.rebuild();
            const tasks: Promise<void>[] = [];
            const nextPool: Rule[] = [];

            for (const i of pool) {
                let found = false;
                for (const o of search.execute(i)) {
                    tasks.push(insertOrIgnore(Fact, o.toString()));
                    count++;
                    if (i.toString() === o.toString()) {
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    nextPool.push(i);
                }
            }
            pool = nextPool;
            await Promise.all(tasks);

            const end = Date.now();
            const duration = (end - begin) / 1000;
            if (count === 0) {
                const delay = Math.max(0, 0.1 - duration);
                await new Promise((resolve) => setTimeout(resolve, delay * 1000));
            }
        }
    } catch (err) {
        // Handle error
    }
}

File: ddss/egg.py
import asyncio
from sqlalchemy import select
from sqlalchemy.ext.asyncio import async_sessionmaker, AsyncSession
from apyds import Rule
from .orm import insert_or_ignore, Facts, Ideas
from .egraph import Search


async def main(session: async_sessionmaker[AsyncSession]) -> None:
    try:
        search = Search()
        pool = []
        max_fact = -1
        max_idea = -1

        while True:
            async with session() as sess:
                for idea in await sess.scalars(select(Ideas).where(Ideas.id > max_idea)):
                    max_idea = max(max_idea, idea.id)
                    pool.append(Rule(idea.data))
                for fact in await sess.scalars(select(Facts).where(Facts.id > max_fact)):
                    max_fact = max(max_fact, fact.id)
                    search.add(Rule(fact.data))
                search.rebuild()
                tasks = []
                next_pool = []
                for i in pool:
                    for o in search.execute(i):
                        tasks.append(asyncio.create_task(insert_or_ignore(sess, Facts, str(o))))
                        if i == o:
                            break
                    else:
                        next_pool.append(i)
                pool = next_pool
                await asyncio.gather(*tasks)
                await sess.commit()

            await asyncio.sleep(0)
    except asyncio.CancelledError:
        pass
